src/Symfony/Component/Process/Process.php

```diff
@@ -1632,6 +1635,38 @@ private function escapeArgument(?string $argument): string
         return '"'.str_replace(['"', '^', '%', '!', "\n"], ['""', '"^^"', '"^%"', '"^!"', '!LF!'], $argument).'"';
     }
 
+    private function replacePlaceholders(string $commandline, array $env)
+    {
+        $isWindows = '\\' === \DIRECTORY_SEPARATOR;
+
+        return preg_replace_callback('/\"\$([_a-zA-Z]++[_a-zA-Z0-9]*)?\"/', function ($matches) use ($isWindows, $commandline, $env) {
+            if (0 === \count($matches)) {
+                return $commandline;
+            }
+            $m = $matches[1];
+
+            if ((!isset($env[$m]) || (isset($env[$m]) && false === $env[$m]))) {
+                foreach ($env as $k => $v) {
+                    if (false === $v) {
+                        unset($env[$k]);
+                    }
+                }
+                if (!$env) {
+                    throw new InvalidArgumentException(sprintf('Invalid command line "%s": no values provided for any placeholders.', $commandline));
+                }
+                $env = implode('", "', array_keys($env));
```

Comment:
I think all the logic to compute the error message is not worth it. I would throw a simple message with the missing key only.